src/Features/Core/Portable/FileHeaders/AbstractFileHeaderDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,98 @@
+                new LocalizableResourceString(nameof(FeaturesResources.A_source_file_is_missing_a_required_header), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+            RoslynDebug.AssertNotNull(DescriptorId);
+
+            var invalidHeaderTitle = new LocalizableResourceString(nameof(FeaturesResources.The_file_header_does_not_match_the_required_text), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+            var invalidHeaderMessage = new LocalizableResourceString(nameof(FeaturesResources.A_source_file_contains_a_header_that_does_not_match_the_required_text), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+            InvalidHeaderDescriptor = CreateDescriptorWithId(DescriptorId, invalidHeaderTitle, invalidHeaderMessage);
+        }
+
+        internal DiagnosticDescriptor MissingHeaderDescriptor => Descriptor;
+
+        internal DiagnosticDescriptor InvalidHeaderDescriptor { get; }
+
+        protected abstract AbstractFileHeaderHelper FileHeaderHelper { get; }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SyntaxTreeWithoutSemanticsAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxTreeAction(HandleSyntaxTree);
+
+        public void HandleSyntaxTree(SyntaxTreeAnalysisContext context)
+        {
+            var tree = context.Tree;
+            var root = tree.GetRoot(context.CancellationToken);
+
+            // don't process empty files
+            if (root.FullSpan.IsEmpty)
+            {
+                return;
+            }
+
+            if (!context.Options.TryGetEditorConfigOption(CodeStyleOptions.FileHeaderTemplate, tree, out string fileHeaderTemplate)
+                || string.IsNullOrEmpty(fileHeaderTemplate))
+            {
+                return;
+            }
+
+            var fileHeader = FileHeaderHelper.ParseFileHeader(root);
+            if (fileHeader.IsMissing)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(MissingHeaderDescriptor, fileHeader.GetLocation(tree)));
+                return;
+            }
+
+            var expectedFileHeader = fileHeaderTemplate.Replace("{fileName}", Path.GetFileName(tree.FilePath));
+            if (!CompareCopyrightText(expectedFileHeader, fileHeader.CopyrightText))
+            {
+                context.ReportDiagnostic(Diagnostic.Create(InvalidHeaderDescriptor, fileHeader.GetLocation(tree)));
+                return;
+            }
+        }
+
+        private static bool CompareCopyrightText(string expectedFileHeader, string copyrightText)
+        {
+            // make sure that both \n and \r\n are accepted from the settings.
+            var reformattedCopyrightTextParts = expectedFileHeader.Replace("\r\n", "\n").Split('\n');
+            var fileHeaderCopyrightTextParts = copyrightText.Replace("\r\n", "\n").Split('\n');
```

Comment:
pattern is in a few places.  maybe have helper method for this?

Reply:
It may not be perfect, but I find this one good enough to leave for now.